helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -722,78 +698,37 @@ public Set<String> getAllInstances() {
   }
 
   /**
-   * Return all the live nodes that are enabled and assignable
-   *
-   * @return A new set contains live instance name and that are marked enabled
-   */
-  public Set<String> getAssignableEnabledLiveInstances() {
-    Set<String> enabledLiveInstances = new HashSet<>(getAssignableLiveInstances().keySet());
-    enabledLiveInstances.removeAll(getDisabledInstances());
-
-    return enabledLiveInstances;
-  }
-
-  /**
-   * Return all the live nodes that are enabled
+   * Return all the live nodes that are enabled. If a node is enabled, it is assignable.
    * @return A new set contains live instance name and that are marked enabled
    */
   public Set<String> getEnabledLiveInstances() {
     Set<String> enabledLiveInstances = new HashSet<>(getLiveInstances().keySet());
-    enabledLiveInstances.removeAll(getDisabledInstances());
+    enabledLiveInstances.retainAll(getEnabledInstances());
 
     return enabledLiveInstances;
   }
 
   /**
-   * Return all nodes that are enabled and assignable.
-   *
-   * @return A new set contains instance name and that are marked enabled
-   */
-  public Set<String> getAssignableEnabledInstances() {
-    Set<String> enabledNodes = new HashSet<>(getAssignableInstances());
-    enabledNodes.removeAll(getDisabledInstances());
-
-    return enabledNodes;
-  }
-
-  /**
-   * Return all nodes that are enabled.
+   * Return all nodes that are enabled. If a node is enabled, it is assignable.
    * @return A new set contains instance name and that are marked enabled
    */
   public Set<String> getEnabledInstances() {
-    Set<String> enabledNodes = new HashSet<>(getAllInstances());
-    enabledNodes.removeAll(getDisabledInstances());
-
-    return enabledNodes;
+    return new HashSet<>(_derivedInstanceCache.getInstanceConfigMapByInstanceOperation(
+        InstanceConstants.InstanceOperation.ENABLE).keySet());
   }
 
   /**
-   * Return all the live nodes that are enabled and assignable and tagged with given instanceTag.
```

Comment:
This cannot be simply removed. We need support backward compatible. The new client can just deprecate all enableInstances API.

But for controller it should understand the enable field for a while for backward compatible. When we remove those deprecated APIs, we can change this logic as well.

Reply:
Discussed offline, backwards compatibility is preserved by HELIX_ENABLED field being respected in getInstanceOperation. If HELIX_ENABLED is set to false, it will have InstanceOperation of DISABLE overriding the InstanceOperation field. If set to true, default to ENABLE InstanceOperation or use set InstanceOperation.